Variable-length `loop_ownership->owned_loop_addrs` is needed for larger programs

Currently, the size of `loop_ownership->owned_loop_addrs` is fixed at 1000, which leads to memory corruption issues when running larger programs that exceed this limit.
Source:
https://github.com/rems-project/cn/blob/09068265fe85169d7cd584328466f60f45001b54/runtime/libcn/src/cn-executable/utils.c#L292-L315

A minimal example to reproduce the issue could look like this:
```c
void myloop(char *p)
/*@
    requires
        take X = each(u64 i; i < 1000u64) { W<char>(array_shift<char>(p, i)) };
    ensures
        take Y = each(u64 i; i < 1000u64) { W<char>(array_shift<char>(p, i)) };
@*/
{

    unsigned long i;
    for (i = 0; i < 1000; i++)
    /*@
        inv take Z = each(u64 j; j < 1000u64) { W<char>(array_shift<char>(p, j)) };
            {p} unchanged;
    @*/
    { }
}

int main(void) {
    char s[1000] = {0};
    myloop(s);
    return 0;
}
```

Though CN can prove the safety of `myloop`, Fulminate failed to terminate successfully due to the memory corruption.

```
> cn verify cn_bug.c
[1/2]: myloop -- pass
other location (File "lib/check.ml", line 1762, characters 40-47)  warning: good/value_check of array of large size: 1000

[2/2]: main -- pass
> cn instrument --run cn_bug.c
./cn_bug.exec.c:151:3: warning: left operand of comma operator has no effect [-Wunused-value]
  151 |   0;
      |   ^
1 warning generated.
************************ Failed at *************************
function myloop, file ./cn_bug.exec.c, line 131
original source location: 
cn_bug.c:24:18:

Loop invariant ownership check failed.
  ==> 0x16b976ae3[0] (0x16b976ae3) not owned
```

Notably, if the constant 1000 is changed to 5, Fulminate terminates correctly.

--- a/runtime/libcn/include/cn-executable/utils.h
+++ b/runtime/libcn/include/cn-executable/utils.h
@@ -148,11 +148,6 @@ typedef struct cn_alloc_id {
 
 typedef hash_table cn_map;
 
-struct loop_ownership {
-  uintptr_t *owned_loop_addrs;
-  signed long arr_size;
-};
-
 void initialise_ownership_ghost_state(void);
 void free_ownership_ghost_state(void);
 void initialise_ghost_stack_depth(void);

--- a/runtime/libcn/src/cn-executable/utils.c
+++ b/runtime/libcn/src/cn-executable/utils.c
@@ -289,29 +289,48 @@ void cn_loop_leak_check(void) {
 
 /* / TODO: DAVID */
 
+typedef struct loop_ownership_nd {
+  uintptr_t addr;
+  size_t size;
+  struct loop_ownership_nd* next;
+} loop_ownership_nd;
+
+struct loop_ownership {
+  loop_ownership_nd* head;
+};
+
+// No destructors: expects to be dropped by the bump allocator _in a timely
+// manner_.
 struct loop_ownership* initialise_loop_ownership_state(void) {
   struct loop_ownership* loop_ownership =
       bump_alloc.malloc(sizeof(struct loop_ownership));
-  loop_ownership->owned_loop_addrs = bump_alloc.malloc(1000 * sizeof(uintptr_t));
-  loop_ownership->arr_size = 0;
+  // assert(loop_ownership);
+  *loop_ownership = (struct loop_ownership){.head = NULL};
   return loop_ownership;
 }
 
+// XXX: Compresses explicit ranges. Catches iteratively fed ranges.
+// To improve: coalesce with any known range, rather than just the last.
 void cn_add_to_loop_ownership_state(
     void* generic_c_ptr, size_t size, struct loop_ownership* loop_ownership) {
-  for (int i = 0; i < size; i++) {
-    loop_ownership->owned_loop_addrs[loop_ownership->arr_size] =
-        (uintptr_t)generic_c_ptr + i;
-    loop_ownership->arr_size++;
+  uintptr_t addr = (uintptr_t)generic_c_ptr;
+  loop_ownership_nd* hd = loop_ownership->head;
+  if (hd && addr == (hd->addr + hd->size))
+    hd->size += size;
+  else {
+    loop_ownership_nd* newhd = bump_alloc.malloc(sizeof(loop_ownership_nd));
+    *newhd = (loop_ownership_nd){.addr = addr, .size = size, .next = hd};
+    loop_ownership->head = newhd;
   }
 }
 
 void cn_loop_put_back_ownership(struct loop_ownership* loop_ownership) {
-  for (int i = 0; i < loop_ownership->arr_size; i++) {
-    int64_t address_key = loop_ownership->owned_loop_addrs[i];
-    ownership_ghost_state_set(&address_key, cn_stack_depth);
+  for (loop_ownership_nd* nd = loop_ownership->head; nd; nd = nd->next) {
+    for (size_t i = 0; i < nd->size; i++) {
+      int64_t addr = nd->addr + i;
+      ownership_ghost_state_set(&addr, cn_stack_depth);
+    }
   }
-  loop_ownership->arr_size = 0;
 }
 
 int ownership_ghost_state_get(int64_t* address_key) {

